Laravel Lambda Queue Worker

adrianspacely opened the pull request:
This PR adds a nice DX for executing queued jobs on lambda.

I am curious to see how many queue worker processes I can run on a single Forge server where the job is processed on lambda. I think what I’ve done should work with Laravel Horizon too which is important for an easy way to get "load test metrics".

**Implementation/Hypothesis:** The queue worker will run on Forge, the job itself executes on Lambda. I.e. Horizon would run on Forge with what I'm hoping can be a large number of workers on a `lambda` queue. All the queue worker is doing is awaiting a tiny sidecar response and managing the queue - the sidecar lambda takes the compute hit.

# To-Do

## Worker

- [x] Extend the queue worker so jobs can process on lambda.
- [x] It works for a job.
- [x] It works for a closure.
- [x] It works for a mailable.
- [x] It works for a notification.
- [x] It works for an event/listeners.
- [x] It works for a custom dispatchable example.
- [x] It releases jobs.
- [x] It releases jobs with a delay.
- [x] It marks jobs as failed via `$this->fail()`.
- [x] It marks jobs as failed via a thrown exception.
- [x] Failed jobs log the thrown exception, not the `LambdaExecutionException`.
- [x] You can dispatch jobs within a lambda job.
- [x] It adds logs from the lambda on the queue worker's server.
- [x] The lambda does not interact with the queue. It returns a payload that includes serialized jobs; exceptions; logs; should fail; should release; and delay for release. _This is convenient because for at least queueing purposes, you don't need to fuss over networking the queue connection to the lambda._
- [x] Add a `LambdaJobProcessing` and a `LambdaJobProcessed` event.
- [x] `ShouldBeUnique` or `ShouldBeUniqueUntilProcessing` will not hold a lock indefinitely. _Confirmed: if the lambda cannot connect to the cache server, it will release until max tries/exceptions are hit at which point it will be marked as failed outside of the lambda which has access to the cache server meaning the lock will release on failure just fine._
- [x] Job chains work and catch callbacks trigger. _(should not happen within the lambda with exception to the callbacks)_
- [x] Job batch will record successful and failed jobs. Then/catch/finally callbacks trigger. It will not double up on counts or failed job UUIDs. _(should not happen within the lambda with exception to the callbacks)_
- [x] Jobs dispatch to the queue via `$this->batch()->add([...])`.

## Feature Toggle

- [x] Add a `RunInLambda` and `DoNotRunInLambda` interface. User will implement this on their jobs.
- [x] `sidecar.queue.enabled`. When `false`, the extended queue worker is never used.
- [x] `sidecar.queue.opt_in_required`. When `true`, you need to implement `RunInLambda` on your job/mailable/etc. When `false`, it will run in lambda unless you implemented `DoNotRunInLambda` on your job.
- [x] `sidecar.queue.allowed_queues` which defaults to `['*']` (all). This allows the user to specify which queue names should allow lambda execution.

## Tests

- [x] Setup Pest PHP.
- [x] Add mock to `PhpLambda` so the tests _(optionally)_ don't have to use Lambda.
- [ ] Setup GitHub actions.
- [x] Source dive how event broadcasting works and see if we need to handle it/cover it with tests.
- [ ] Ensure this works with Laravel Horizon; load test; and share the Horizon metrics here to see how many processes can be run on a server.

**Other Notes:**

I made my decisions around these opinions:

- **the user should not have to network the queue connection to the lambdas to use this feature.**
- **logs should go to their configured log channel, not cloud watch.**

Some additional notes:

- I delegate all queue management to the queue worker with the payload I return. The lambda does not require access to the redis/sqs/beanstalkd/whatever queue connection.
- Use case for `LambdaJobProcessing` and `LambdaJobProcessed` events: We can suggest in the docs that the user add a listener to emit `JobProcessing` and `JobProcessed` events so that the before/after job behaviour also happens in the lambda. _For example, [spatie/laravel-multitenancy](https://github.com/spatie/laravel-multitenancy/blob/ae44dd12e9f244c43ed04479e7f5f85c9320e2ec/src/Actions/MakeQueueTenantAwareAction.php#L41) sets the current tenant before running a job, so in this case, you would want the same to happen in the lambda. I don't want to make the assumption that the user wants to run something twice._
- You can add jobs to a batch dynamically with `$this->batch()->add([...])`. We catch those added jobs within the lambda and return them for dispatch by the worker just like we do with any dispatched job. The `then`, `catch`, and `finally` batch callbacks run within the lambda. The batch totals, however, increment within the lambda - the reason being is that you must be able to fetch the batch within the lambda, therefore, you would be able to write to the batch totals. I imagine if you're using job batches, then you would have configured your lambdas to be able to connect to the database.

adrianspacely committed:
Added a lambda queue worker which decorates to job to fire in the lambda.

adrianspacely committed:
Lambda queue worker now triggers with opt-in config and support for RunOnLambda and DoNotRunInLambda interfaces.

adrianspacely committed:
Delete useless job getter.

adrianspacely committed:
Only serialise $job in the closure.

adrianspacely committed:
Fire job closure now serialises for Lambda.

adrianspacely committed:
Keep things simple on lambda by using a SyncJob when calling the queue handler.

adrianspacely committed:
Delete/release the job if it was done in the SyncJob on the Lambda.

adrianspacely committed:
Move invade to the decorator helper.

adrianspacely committed:
Add config for allowed queues.

aarondfrancis commented:
This is madness and I couldn't love it more. I've just given it a cursory look so far, but I'm going to dive in more this week. 

There are a few things I need to catch up on, mainly bringing Sidecar up to speed with Laravel 9 and potentially PHP 8.1 if there's anything missing there. I'm gonna get those things rolling and then come back to this. 

Stefan did get real testing sorted out, so I'll implement that as well. The whole "mock everything" method of testing that I've been doing is not great. It feels so brittle.

Again, this totally rules and I'm pumped to try it out!

adrianspacely committed:
Test stubs.

adrianspacely committed:
Move the extended SyncJob to a single-use anonymous object.

adrianb93 commented:
@aarondfrancis, would it be ok if I used [Pest](https://pestphp.com/) for the tests? The data providers would reduce the number of test functions I need to implement. It also doesn't prevent you from using regular PHPUnit in other test files if you prefer that.

aarondfrancis commented:
Yup go for it!

On Sun, Feb 13, 2022 at 6:23 PM Adrian Brown ***@***.***>
wrote:

> @aarondfrancis <https://github.com/aarondfrancis>, would it be ok if I
> used Pest <https://pestphp.com/> for the tests? The data providers would
> reduce the number of test functions I need to implement. It also doesn't
> prevent you from using regular PHPUnit in other test files if you prefer
> that.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hammerstonedev/sidecar-php/pull/1#issuecomment-1038493182>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGXKCYH7JBTRAO3EC24CXTU3BDRPANCNFSM5OIAJZUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

adrianspacely committed:
Add test queuable classes.

adrianspacely committed:
Config helpers.

adrianspacely committed:
tests

adrianspacely committed:
Lambda listeners for test assertions.

adrianspacely committed:
Working on tests and the worker throws the same exception from the lambda for better failed_jobs and logs.

adrianb93 commented:
@aarondfrancis, it might be my own dumb fault - so I'll keep looking into it. Out of curiosity, does your Laracon demo still work for you?

I keep getting 5-second timeouts on Lambda from `ap-southeast-2` even for something simple like `artisan inspire`. I've been mocking in order to move forward on this PR.

adrianspacely committed:
Implement mark as failed tests and cleanup the datasets

adrianspacely committed:
Cleanup test descriptions.

adrianspacely committed:
Add events/listeners and queued classes which don't interact with the queue for release/fail to the datasets.

adrianspacely committed:
Add mocking to PhpLambda (invokes the closure in the same process)

adrianspacely committed:
Mock if it has been configured.

adrianspacely committed:
Add LambdaJobProcessing and LambdaJobProcessed.

adrianspacely committed:
Checkbox for equal length dataset descriptions.

adrianspacely committed:
Can now dispatch jobs within the lambda.

adrianspacely committed:
Cleanup

adrianspacely committed:
Test that chained jobs work including where the catch callback is handled.

adrianspacely committed:
Forgot throw

adrianspacely committed:
Test stubs for job batches.

adrianspacely committed:
Job batch tests.

adrianspacely committed:
Support for dynamically adding jobs to a batch.

adrianspacely committed:
Log handling.

adrianspacely committed:
Include package vendor for tests.

adrianspacely committed:
Add package files (should autoload with psr4)

adrianspacely committed:
Lambda doesn't like the testbench bootstrap.

adrianspacely committed:
Fix bootstrap kernels

adrianspacely committed:
Set all app cache paths.

adrianspacely committed:
Autoload the test support files on the lambda.

adrianspacely committed:
Autoload test support files.

adrianspacely committed:
Storage step and include the .env for tests

adrianspacely committed:
Unsuccessful attempt to load pest test classes so that queued closures are available.

adrianspacely committed:
Get the pest classes included on the Lambda for testing queued closures.

The next plan would have been serialised closures.

adrianspacely committed:
Use the log for the batch/chain tests.

There appears to be a bug where the lambda is handling failed jobs and batches instead of the worker.
